In [5]:
import pandas as pd
import numpy as np
df=pd.read_csv('/Users/jussaragaspar/Downloads/heart.csv')

In [6]:
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [7]:
#model 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
X = df.select_dtypes(include=['int64','float64'])
y = df['HeartDisease']



In [8]:
print(X.dtypes)
print(X.isnull().sum())


Age               int64
RestingBP         int64
Cholesterol       int64
FastingBS         int64
MaxHR             int64
Oldpeak         float64
HeartDisease      int64
dtype: object
Age             0
RestingBP       0
Cholesterol     0
FastingBS       0
MaxHR           0
Oldpeak         0
HeartDisease    0
dtype: int64


In [9]:
train_X,test_X,train_y,test_y=train_test_split(X,y,test_size=0.2,random_state=42)
model=LogisticRegression()

from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(train_X, train_y)
y_pred = knn.predict(test_X)
accuracy = accuracy_score(test_y, y_pred)
print(f"Accuracy: {accuracy:.2f}") 

Accuracy: 0.71


In [11]:
from sklearn.metrics import classification_report, roc_auc_score, average_precision_score
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(train_X, train_y)
# Class predictions
y_pred = knn.predict(test_X)

# Probabilities for positive class (class 1)
y_proba = knn.predict_proba(test_X)[:, 1]
print(classification_report(test_y, y_pred))

              precision    recall  f1-score   support

           0       0.63      0.71      0.67        77
           1       0.77      0.70      0.74       107

    accuracy                           0.71       184
   macro avg       0.70      0.71      0.70       184
weighted avg       0.71      0.71      0.71       184



Accuracy = 0.71
The model correctly predicts 71% of patients overall.
That’s decent, not amazing, not bad.

Precision (Class 1) = 0.77

When the model predicts Heart Disease,
it is correct 77% of the time.
That’s good.

 Recall (Class 1) = 0.70
The model detects 70% of actual heart disease cases.

This means:
 It misses 30% of sick patients.
In healthcare, recall is VERY important.

In [12]:
roc_score = roc_auc_score(test_y, y_proba)
print("ROC AUC:", roc_score)
avg_precision = average_precision_score(test_y, y_proba)
print("Average Precision Score:", avg_precision)


ROC AUC: 0.7299429542420197
Average Precision Score: 0.7484664959939336


ROC AUC
1 patient with heart disease

1 patient without

The model will rank the sick patient higher 73% of the time.